test.js

```diff
@@ -69,10 +95,106 @@ test('it supports multiple arguments', t => {
   source.emit(1, 'qu');
 
   t.deepEqual(
-    sink.getReceivedData(), 
+    sink.getReceivedData(),
     ['foo','bar','baz','qu'],
     'sink gets seed and subsequent data'
   );
 
   t.end();
 });
+
+test('it queues sync completion', t => {
+  const seededSrc = startWith('a', 'b', 'c')(of('d'));
+
+  const sink = makeMockCallbag();
+
+  seededSrc(0, sink);
+
+  t.deepEqual(
+    sink.getMessages().slice(1),
+    [
+      [1, 'a'],
+      [1, 'b'],
+      [1, 'c'],
+      [1, 'd'],
+      [2, undefined],
+    ],
+    'sink gets data in the correct order'
+  );
+
+  t.end();
+});
+
+test('it doesn\'t request data when receiving uknown type', t => {
+  const source = makeMockCallbag(true);
+  const seedWithFoo = startWith('foo');
+  const seededSource = seedWithFoo(source);
+
+  const autoPull = () => {};
+  forEach(autoPull)(seededSource);
+
+  source.emit(1, 'a');
+  source.emit(1, 'b');
+  source.emit('unknown', 'c');
+  source.emit(1, 'd');
+  source.emit('unknown', 'e');
+
+  t.deepEqual(
+    source.getMessages().slice(1),
+    [
+      [1, undefined], // request sent up on 0
+      [1, undefined], // request sent up startWith seed
+      [1, undefined],
+      [1, undefined],
+      [1, undefined],
+    ],
+    'source gets correct number of requests from sink'
+  );
+
+  t.end();
+});
+
+test('it passes sink errors up (& data for unknown types too)', t => {
+  const source = makeMockCallbag(true);
+  const seedWithFoo = startWith('foo');
+  const sink = makeMockCallbag();
+
+  seedWithFoo(source)(0, sink);
+
+  sink.emit(1);
+  sink.emit('unknown', 'payload');
+  sink.emit(11, 'other_data');
+  sink.emit(2, 'err');
+
+  t.deepEqual(
+    source.getMessages().slice(1),
+    [
+      [1, undefined],
+      ['unknown', 'payload'],
+      [11, 'other_data'],
+      [2, 'err'],
+    ],
+    'source gets type & data from sink'
+  );
+
+  t.end();
+});
+
+test('it stops emitting after receiving unsubscription request', t => {
+  let history = [];
+
+  const seededSrc = startWith('a', 'b', 'c')(fromIter(['d']));
+  const limitedSrc = take(2)(seededSrc);
+
+  const sink = makeMockCallbag();
+
+  limitedSrc(0, sink);
```

Comment:
We're getting a fair few steps in the chain now. Do you think perhaps it is worth to add `callbag-pipe` as a devDep and rephase as 👇 ?

```javascript
const sink = makeMockCallbag();
pipe(
  fromIter(['d']),
  startWith('a', 'b', 'c'),
  take(2),
  sink
);
```

Reply:
👍 Ive kinda wanted to write it this way but u werent using `pipe` in any other test, so I've decided against it